File: browser_app/crypto_chacha20.js
// --- 2. CHA-CHA20 CORE (Sesuai RFC 8439) ---
    
function rotl(x, n) {
  return ((x << n) | (x >>> (32 - n))) >>> 0;
}

function quarterRound(s, a, b, c, d) {
  s[a] = (s[a] + s[b]) >>> 0; s[d] = rotl(s[d] ^ s[a], 16);
  s[c] = (s[c] + s[d]) >>> 0; s[b] = rotl(s[b] ^ s[c], 12);
  s[a] = (s[a] + s[b]) >>> 0; s[d] = rotl(s[d] ^ s[a], 8);
  s[c] = (s[c] + s[d]) >>> 0; s[b] = rotl(s[b] ^ s[c], 7);
}

function chacha20Block(key, nonce, counter = 0) { 
  const s = new Uint32Array(16);
  
  // 1. Constants (Little-Endian)
  s[0] = 0x61707865; s[1] = 0x3320646e; s[2] = 0x79622d32; s[3] = 0x6b206574;
  
  // 2. Key (32 bytes, Little-Endian)
  for (let i = 0; i < 8; i++) {
      const offset = i * 4;
      s[4 + i] = key[offset] 
               | (key[offset + 1] << 8) 
               | (key[offset + 2] << 16) 
               | (key[offset + 3] << 24);
  }
  
  // 3. Counter (4 bytes, Little-Endian)
  s[12] = counter >>> 0; 
      
  // 4. Nonce/IV (12 bytes, Little-Endian)
  for (let i = 0; i < 3; i++) {
      const offset = i * 4;
      s[13 + i] = nonce[offset] 
                | (nonce[offset + 1] << 8) 
                | (nonce[offset + 2] << 16) 
                | (nonce[offset + 3] << 24);
  }
  
  const working = new Uint32Array(s);
  
  for (let i = 0; i < 10; i++) {
      // Column Rounds
      quarterRound(working, 0, 4, 8, 12);
      quarterRound(working, 1, 5, 9, 13);
      quarterRound(working, 2, 6, 10, 14);
      quarterRound(working, 3, 7, 11, 15);
      // Diagonal Rounds
      quarterRound(working, 0, 5, 10, 15);
      quarterRound(working, 1, 6, 11, 12);
      quarterRound(working, 2, 7, 8, 13);
      quarterRound(working, 3, 4, 9, 14);
  }

  const out = new Uint8Array(64);
  for (let i = 0; i < 16; i++) {
      const v = (working[i] + s[i]) >>> 0;
      // Konversi 32-bit word ke 4 byte Little-Endian
      out[i * 4] = v & 0xff;
      out[i * 4 + 1] = (v >>> 8) & 0xff;
      out[i * 4 + 2] = (v >>> 16) & 0xff;
      out[i * 4 + 3] = (v >>> 24) & 0xff;
  }
  return out;
}

// ==================== ChaCha20 XOR ====================
export function chacha20XOR(inputBytes, key, nonce, initialCounter = 1) {
  if (key.length !== 32 || nonce.length !== 12) {
      throw new Error("Key harus 32 byte dan Nonce harus 12 byte (RFC 8439).");
  }
  if (!(inputBytes instanceof Uint8Array)) {
      throw new Error("Input harus Uint8Array.");
  }

  const out = new Uint8Array(inputBytes.length);
  let blockCount = initialCounter; 
  
  for (let pos = 0; pos < inputBytes.length; pos += 64) {
      const block = chacha20Block(key, nonce, blockCount);
      for (let i = 0; i < 64 && pos + i < inputBytes.length; i++) {
          out[pos + i] = inputBytes[pos + i] ^ block[i];
      }
      blockCount++;
  }
  return out;
}

File: server_python/crypto_chacha20.py
import struct

# Fungsi utilitas rotasi 32-bit ke kiri
def rotl(x, n):
    """Rotate Left 32-bit."""
    # (x << n) | (x >>> (32 - n))
    return ((x << n) & 0xFFFFFFFF) | (x >> (32 - n))

# Fungsi Quarter Round
def quarter_round(x, a, b, c, d):
    """ChaCha20 Quarter Round operation on indices a, b, c, d."""
    # (x[a] + x[b])
    x[a] = (x[a] + x[b]) & 0xFFFFFFFF
    # (x[d] ^ x[a]), rotl 16
    x[d] = rotl(x[d] ^ x[a], 16)

    # (x[c] + x[d])
    x[c] = (x[c] + x[d]) & 0xFFFFFFFF
    # (x[b] ^ x[c]), rotl 12
    x[b] = rotl(x[b] ^ x[c], 12)

    # (x[a] + x[b])
    x[a] = (x[a] + x[b]) & 0xFFFFFFFF
    # (x[d] ^ x[a]), rotl 8
    x[d] = rotl(x[d] ^ x[a], 8)

    # (x[c] + x[d])
    x[c] = (x[c] + x[d]) & 0xFFFFFFFF
    # (x[b] ^ x[c]), rotl 7
    x[b] = rotl(x[b] ^ x[c], 7)

# Fungsi Inisialisasi State dan Komputasi Block
def chacha20_block(key_bytes, nonce_bytes, counter):
    """
    Menghasilkan satu keystream block (64 byte) ChaCha20.
    Key 32B, Nonce 12B, Counter 4B.
    """
    # 1. Inisialisasi State Awal (16 words 32-bit)
    state = [0] * 16

    # Constants (s[0]-s[3]): "expa", "nd 32-", "byte", " k"
    # Format 'I' adalah unsigned 32-bit integer, '<' adalah Little-Endian
    constants = struct.unpack('<4I', b'expand 32-byte k')
    state[0:4] = list(constants)

    # Key (s[4]-s[11]): 32 bytes (Little-Endian)
    state[4:12] = struct.unpack('<8I', key_bytes)
    
    # Counter (s[12]): 4 bytes (Little-Endian)
    state[12] = counter & 0xFFFFFFFF
    
    # Nonce/IV (s[13]-s[15]): 12 bytes (Little-Endian)
    state[13:16] = struct.unpack('<3I', nonce_bytes)

    # State kerja untuk 20 putaran
    working_state = list(state)
    
    # 2. Komputasi Block (20 Putaran)
    for _ in range(10): # 10 iterasi = 20 putaran
        # Column Rounds
        quarter_round(working_state, 0, 4, 8, 12)
        quarter_round(working_state, 1, 5, 9, 13)
        quarter_round(working_state, 2, 6, 10, 14)
        quarter_round(working_state, 3, 7, 11, 15)

        # Diagonal Rounds
        quarter_round(working_state, 0, 5, 10, 15)
        quarter_round(working_state, 1, 6, 11, 12)
        quarter_round(working_state, 2, 7, 8, 13)
        quarter_round(working_state, 3, 4, 9, 14)

    # 3. Final Summation dan Output
    output_words = [
        (working_state[i] + state[i]) & 0xFFFFFFFF
        for i in range(16)
    ]
    
    # Konversi 16 words 32-bit (Little-Endian) kembali ke 64 byte
    # '<16I' artinya 16 unsigned 32-bit integers, Little-Endian
    return struct.pack('<16I', *output_words)

# Fungsi Enkripsi/Dekripsi Utama (XOR)
def chacha20_xor(data_bytes, key_bytes, nonce_bytes, initial_counter=1):
    """
    Melakukan enkripsi atau dekripsi XOR dengan ChaCha20.
    Data harus berupa bytes, Key 32B, Nonce 12B.
    """
    
    if len(key_bytes) != 32:
        raise ValueError("Key harus 32 byte (256-bit).")
    if len(nonce_bytes) != 12:
        raise ValueError("Nonce harus 12 byte (96-bit), sesuai RFC 8439.")
    
    data_length = len(data_bytes)
    output_bytes = bytearray(data_length)
    block_size = 64
    
    block_counter = initial_counter
    
    for offset in range(0, data_length, block_size):
        # 1. Generate Keystream Block
        keystream_block = chacha20_block(key_bytes, nonce_bytes, block_counter)
        
        # Tentukan panjang data yang tersisa di blok ini
        length_to_xor = min(block_size, data_length - offset)
        
        # 2. XOR dengan Data
        for i in range(length_to_xor):
            output_bytes[offset + i] = data_bytes[offset + i] ^ keystream_block[i]
            
        block_counter += 1
        
    return bytes(output_bytes)

def chacha20_encrypt(plaintext: bytes, key: bytes, nonce: bytes) -> bytes:
    cipher= chacha20_xor(plaintext, key, nonce, initial_counter=1)
    return cipher

def chacha20_decrypt(plaintext: bytes, key: bytes, nonce: bytes) -> bytes:
    plain = chacha20_xor(plaintext, key, nonce, initial_counter=1)
    return plain